tests/internals/cache_metrics/cache_metrics_test.go

```diff
@@ -0,0 +1,212 @@
+          image: nginx
+`
+
+	scaledObjectTemplate = `
+apiVersion: keda.sh/v1alpha1
+kind: ScaledObject
+metadata:
+  name: {{.ScaledObjectName}}
+  namespace: {{.TestNamespace}}
+spec:
+  advanced:
+    horizontalPodAutoscalerConfig:
+      behavior:
+        scaleDown:
+          stabilizationWindowSeconds: 1
+  scaleTargetRef:
+    name: {{.DeploymentName}}
+  pollingInterval: 1000
+  minReplicaCount: 0
+  maxReplicaCount: 10
+  cooldownPeriod: 0
+  triggers:
+    - type: kubernetes-workload
+      useCachedMetrics: {{.EnableUseCachedMetrics}}
+      metadata:
+        podSelector: 'app={{.MonitoredDeploymentName}}'
+        value: '2'
+`
+)
+
+func TestScaler(t *testing.T) {
+	// setup
+	t.Log("--- setting up ---")
+	// Create kubernetes resources
+	kc := GetKubernetesClient(t)
+	data, templates := getTemplateData()
+
+	CreateKubernetesResources(t, kc, testNamespace, data, templates)
+
+	// test direct metrics query (the standard behavior)
+	testDirectQuery(t, kc, data)
+
+	// test querying metrics on polling interval
+	testCacheMetricsOnPollingInterval(t, kc, data)
+
+	// cleanup
+	DeleteKubernetesResources(t, kc, testNamespace, data, templates)
+}
+
+func getTemplateData() (templateData, []Template) {
+	return templateData{
+			TestNamespace:               testNamespace,
+			DeploymentName:              deploymentName,
+			ScaledObjectName:            scaledObjectName,
+			MonitoredDeploymentName:     monitoredDeploymentName,
+			MonitoredDeploymentReplicas: defaultMonitoredDeploymentReplicas,
+		}, []Template{
+			{Name: "deploymentTemplate", Config: deploymentTemplate},
+			{Name: "monitoredDeploymentTemplate", Config: monitoredDeploymentTemplate},
+		}
+}
+
+func testCacheMetricsOnPollingInterval(t *testing.T, kc *kubernetes.Clientset, data templateData) {
+	t.Log("--- testing caching metrics on polling interval ---")
+
+	KubectlApplyWithTemplate(t, data, "deploymentTemplate", deploymentTemplate)
+	data.MonitoredDeploymentReplicas = defaultMonitoredDeploymentReplicas
+	KubectlApplyWithTemplate(t, data, "monitoredDeploymentTemplate", monitoredDeploymentTemplate)
+
+	// initial replica count for a deployment
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 1, 60, 1),
+		"replica count should be 1 after 1 minute")
+
+	// initial replica count for a monitored deployment
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, defaultMonitoredDeploymentReplicas, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	data.EnableUseCachedMetrics = true
+	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)
+
+	// Metric Value = 8, DesiredAverageMetricValue = 2
+	// should scale in to 8/2 = 4 replicas, irrespective of current replicas
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 4, 60, 1),
+		"replica count should be 4 after 1 minute")
```

Comment:
I'd give more time, something like 2 minutes or 3, this helper will finish once the target replica count is reached, and 1 minute having to scale to 4, could not be enough time

Reply:
@zroubalik Is this resolved?